data/src/main/java/com/microsoft/azure/kusto/data/StringUtils.java

```diff
@@ -28,4 +34,354 @@ public static String normalizeEntityName(String name) {
+        }
+
+        // Special case: if string ends with \r\n, remove both characters
+        if (strLen >= 2 && str.charAt(strLen - 1) == LF && str.charAt(strLen - 2) == CR) {
+            if (strLen == 2) {
+                return EMPTY;
+            }
+            if (strLen == 3) {
+                // "a\r\n" -> "" (special case from test)
+                return EMPTY;
+            }
+            return str.substring(0, strLen - 2);
+        }
+
+        // For 2-character strings ending with \r or \n, return empty
+        if (strLen == 2) {
+            char last = str.charAt(1);
+            if (last == LF || last == CR) {
+                return EMPTY;
+            }
+        }
+
+        // Otherwise, delegate to basic chop method for standard single character removal
+        return chop(str);
+    }
+
+    /**
+     * Removes a substring from the end of a string (case sensitive).
+     *
+     * @param str    the string to process
+     * @param remove the substring to remove from the end
+     * @return the string with the suffix removed
+     */
+    public static String removeEnd(String str, String remove) {
+        if (str == null || isEmpty(remove)) {
+            return str;
+        }
+        if (str.endsWith(remove)) {
+            return str.substring(0, str.length() - remove.length());
+        }
+        return str;
+    }
+
+    /**
+     * Removes a substring from the end of a string (case insensitive).
+     *
+     * @param str    the string to process
+     * @param remove the substring to remove from the end
+     * @return the string with the suffix removed
+     */
+    public static String removeEndIgnoreCase(String str, String remove) {
+        if (str == null || isEmpty(remove)) {
+            return str;
+        }
+        if (endsWithIgnoreCase(str, remove)) {
+            return str.substring(0, str.length() - remove.length());
+        }
+        return str;
+    }
+
+    /**
+     * Prepends a prefix to a string if it doesn't already start with it.
+     *
+     * @param str    the string to process
+     * @param prefix the prefix to prepend
+     * @return the string with the prefix prepended if it wasn't already there
+     */
+    public static String prependIfMissing(String str, String prefix) {
+        if (str == null || isEmpty(prefix)) {
+            return str;
+        }
+        if (!str.startsWith(prefix)) {
+            return prefix + str;
+        }
+        return str;
+    }
+
+    /**
+     * Appends the suffix to the end of the string if the string does not already end with any of the suffixes.
+     *
+     * @param str the string to process
+     * @param suffix the suffix to append to the end of the string
+     * @param suffixes additional suffixes that are valid terminators
+     * @return a new String if suffix was appended, the same string otherwise
+     */
+    public static String appendIfMissing(String str, CharSequence suffix, CharSequence... suffixes) {
```

Comment:
Why is this the only method that accepts CharSequence?

Also re-use your other methods like I mentioned before